### DISCUSSION TITLE
v11 adds blur to bitmap render where v0.10 shows sharp and original pixels

### DISCUSSION THREAD

**Discussion by sn4k3** [AvaloniaUI/Avalonia#12060] [Q&A] [answered]
I'm migrating from v0.10 to v11, but my custom control which draws a bitmap is adding a blur on v11. How to get sharp pixels instead?

Left: v0.10
Right: v11

![image](https://github.com/AvaloniaUI/Avalonia/assets/113281/d05648cb-4f95-47c1-abc7-f93ecf2aefbc)

The render method (untouched):

```C#
public override void Render(DrawingContext context)
    {
        //Debug.WriteLine($"Render: {DateTime.Now.Ticks}");
        base.Render(context);
        
        var viewPortSize = ViewPortSize;
        // Draw Grid
        var gridCellSize = GridCellSize;
        if (ShowGrid & gridCellSize > 0 && (!IsHorizontalBarVisible || !IsVerticalBarVisible))
        {
            // draw the background
            var gridColor = GridColor;
            var altColor = GridColorAlternate;
            var currentColor = gridColor;
            for (int y = 0; y < viewPortSize.Height; y += gridCellSize)
            {
                var firstRowColor = currentColor;

                for (int x = 0; x < viewPortSize.Width; x += gridCellSize)
                {
                    context.FillRectangle(currentColor, new Rect(x, y, gridCellSize, gridCellSize));
                    currentColor = ReferenceEquals(currentColor, gridColor) ? altColor : gridColor;
                }

                if (Equals(firstRowColor, currentColor))
                    currentColor = ReferenceEquals(currentColor, gridColor) ? altColor : gridColor;
            }

        }
        /*else
        {
            context.FillRectangle(Background, new Rect(0, 0, Viewport.Width, Viewport.Height));
        }*/

        var image = Image;
        if (image is null) return;
        var imageViewPort = GetImageViewPort();


        // Draw iamge
        context.DrawImage(image,
            GetSourceImageRegion(),
            imageViewPort
        );

        var zoomFactor = ZoomFactor;
        
        if (HaveTrackerImage && _pointerPosition is {X: >= 0, Y: >= 0})
        {
            var destSize = TrackerImageAutoZoom
                ? new Size(_trackerImage!.Size.Width * zoomFactor, _trackerImage.Size.Height * zoomFactor)
                : image.Size;

            var destPos = new Point(
                _pointerPosition.X - destSize.Width / 2,
                _pointerPosition.Y - destSize.Height / 2
            );
            context.DrawImage(_trackerImage,
                new Rect(destPos, destSize)
            );
        }

        //SkiaContext.SkCanvas.dr
        // Draw pixel grid
        if (zoomFactor > PixelGridZoomThreshold && SizeMode == SizeModes.Normal)
        {
            var offsetX = Offset.X % zoomFactor;
            var offsetY = Offset.Y % zoomFactor;

            Pen pen = new(PixelGridColor);
            for (double x = imageViewPort.X + zoomFactor - offsetX; x < imageViewPort.Right; x += zoomFactor)
            {
                context.DrawLine(pen, new Point(x, imageViewPort.X), new Point(x, imageViewPort.Bottom));
            }

            for (double y = imageViewPort.Y + zoomFactor - offsetY; y < imageViewPort.Bottom; y += zoomFactor)
            {
                context.DrawLine(pen, new Point(imageViewPort.Y, y), new Point(imageViewPort.Right, y));
            }

            context.DrawRectangle(pen, imageViewPort);
        }

        if (SelectionRegion != default)
        {
            var rect = GetOffsetRectangle(SelectionRegion);
            var selectionColor = SelectionColor;
            context.FillRectangle(selectionColor, rect);
            var color = Color.FromArgb(255, selectionColor.Color.R, selectionColor.Color.G, selectionColor.Color.B);
            context.DrawRectangle(new Pen(color.ToUint32()), rect);
        }
    }
```

**Comment by maxkatz6** [comment, marked as answer]
Can you check if RenderOptions.BitmapInterpolationMode = None (or try other RenderOptions) help here. Try setting it on parent control, since you can't send it to the drawing context.

**Reply by sn4k3** [reply]
That worked...
I set `RenderOptions.SetBitmapInterpolationMode(this, BitmapInterpolationMode.None);` on control constructor.
Thanks

**Reply by fitdev** [reply]
Just a suggestion. I stumbled across this answer after 1 hour of searching. It would be nice if `RenderOptions` are mentioned in your new [docs on graphics](https://docs.avaloniaui.net/docs/next/guides/graphics-and-animation/graphics-and-animations) and in the custom control section too.

**Reply by fitdev** [reply]
And, as an aside, it would be also great to know (perhaps this can be added to the docs too), why this design was chosen, i.e. a struct with a static method, as opposed to having appropriate instance members on `DrawingContext` as is the case with `System.Drawing.Graphics`? The latter seems to be far more discoverable.

**Reply by fitdev** [reply]
Sorry for yet another comment, I noticed that despite me having `RenderOptions.SetEdgeMode(this, EdgeMode.Aliased);` within `public override void Render(DrawingContext context)`, when the control is first displayed it is rendered with `EdgeMode.Antialias` (looks blurry), and _only when it gets focus_ does it render properly (`EdgeMode.Aliased`) and stays that way. Why is it not rendered with `EdgeMode.Aliased` from the get go?

**Reply by Gillibald** [reply]
Because you set it after the first render happened....

**Reply by fitdev** [reply]
@Gillibald Don't think so. I am calling it in the overridden `Render` as the first line.

```cs
  public override void Render(DrawingContext context) {
    RenderOptions.SetEdgeMode(this, EdgeMode.Aliased); //Very first call in my custom control
    //Rest of render logic here
  }
```

**Reply by Gillibald** [reply]
RenderOptions are no property of the DrawingContext. They are a property of the visual. So the moment Render is called RenderOptions for that render pass are already defined. You change the Visual's RenderOptions for the next render.

**Reply by fitdev** [reply]
I am not sure I understand.

1. I set RenderOptions as the _very first_ thing in my Render method. It would not render anything unless it has called into my Render method. But because it did render something (i.e. ran my Render code) it _must have_ called `RenderOptions.SetEdgeMode(this, EdgeMode.Aliased);` because it is the _very first line_ in that method. So, I do not understand how is it not setting the render options if it is literally the first thing in the render method, and no rendering can take place unless Render method is called.
2. And then if I am not supposed to set it in the Render method, then where should I set this? In the constructor?

**Reply by fitdev** [reply]
Update: Calling `RenderOptions.SetEdgeMode(this, EdgeMode.Aliased);`  from the constructor made the problem go away. But I still would like to know why it does not work if I do it as the very first thing in my completely overridden Render method. I mean how can it render anything at all without first executing that very same line that sets RenderOptions?
